File: 프로그래머스/알고리즘 고득점 Kit/기능개발.java
import java.util.*;

class Solution {
	public int[] solution(int[] progresses, int[] speeds) {
		List<Integer> list = new ArrayList<>();
		int index = 0;

		while(index < progresses.length) {
			final int cal = (100 - progresses[index]) / speeds[index];
			final int need = (100 - progresses[index]) % speeds[index] == 0 ? cal : cal + 1;

			for (int i=index ; i<progresses.length ; i++) {
				progresses[i] += need * speeds[i];
			}

			int count = 0;
			for (int i=index ; i<progresses.length ; i++) {
				if (progresses[i] < 100) {
					index = i;
					break;
				}

				if (i == progresses.length - 1) {
					index = i + 1;
				}
				count++;
			}

			list.add(count);
		}

		return list.stream().mapToInt(Integer::intValue).toArray();
	}
}

File: 프로그래머스/알고리즘 고득점 Kit/기능개발.kt
import java.util.*
import java.util.stream.IntStream

class Solution {

    fun solution(progresses: IntArray, speeds: IntArray): IntArray {
        val list: MutableList<Int> = ArrayList()
        var index = 0

        while (index < progresses.size) {
            val cal = (100 - progresses[index]) / speeds[index]
            val need = if ((100 - progresses[index]) % speeds[index] == 0) cal else cal + 1

            IntStream.range(index, progresses.size).forEach { idx: Int -> progresses[idx] += need * speeds[idx] }

            var count = 0
            for (i in index until progresses.size) {
                when {
                    progresses[i] < 100 -> {
                        index = i
                        break
                    }
                    i == progresses.size - 1 -> {
                        index = i + 1
                        count++
                    }
                    else -> count++
                }
            }

            list.add(count)
        }

        return list.toIntArray()
    }
}
